lib/pbench/server/api/resources/__init__.py

```diff
@@ -1460,7 +1460,9 @@ def _get_dataset_metadata(
                 if native_key == Metadata.USER:
                     user = Auth.token_auth.current_user()
                 try:
-                    metadata[i] = Metadata.getvalue(dataset=dataset, key=i, user=user)
+                    metadata[i] = Metadata.getvalue(
+                        dataset=dataset, key=i, user=user.username if user else None
+                    )
```

Comment:
The only real difference is that translation between ID and name belongs to Keycloak; or, more generically, our external OIDC provider, instead of with an internal `User` SQL table. And that means an authoritative translation requires an external call.

This is one reason I've always assumed we want to *require* and *depend on* a closely associated Keycloak broker. If a user has been logged in through the broker, it should have live cached data on the registration and attributes of that user, so the translations will be quick. (It's not like our SQL translations are really "local" after all, since we rely on an external PostgreSQL server: we're just moving our "`User` DB" from the PostgreSQL backend to the Keycloak backend. Which is a bit of a simplification as our `User` lookups through `Dataset` and `Metadata` came from a single SQL join rather than two separate network transactions, but pretend I didn't say that; and a join isn't "free" either.)

I would prefer not to try to cache translations locally, because that just buys into synchronization issues we don't want to own. And even if we *did* have a local cache, scattering this information across various tables doesn't make sense: we'd have a `UserCache` table (which pragmatically means we're back to using our PostgreSQL backend instead of our Keycloak backend.)

__And yes, internally we should only be using the Keycloak broker's user UUID as our "foreign key" in user-related tables.__

Reply:
I agree, We need to use the UUID for the user identification, we can not use the username as its mutable.

Reply:
> Still, there is an interesting problem: what should the Pbench Server API take in order to specify a "user"? If it is a UUID, then we put the load on the client to convert back and forth between UUIDs and user names, which is likely to be a heavy lift. If we do the conversion on the Server side, we have to keep in mind that a "user name" is not necessarily a unique identifier. This is OK in many contexts (e.g., searching for results); however, is it _**not**_ OK in other contexts, such as selecting users to add to groups and roles. So, we're going to need to keep our wits about us and we carry this stuff forward.

I think the server do need to contact the identity broker if we want to get the most up-to-date user identification information (as we are not storing any user information in an internal table). It is yet to be seen how computation-heavy that transaction would be compared to an internal database query. But the service client or admin token can be used to get any information about any user that our broker stores. However, w.r.t this PR I think that would be out of the scope to address it here. But that's something we definitely need to discuss going forward with the Keycloak integration.

Reply:
> I think the server do[es] need to contact the identity broker if we want to get the most up-to-date user identification information

Yes, but...I don't think we need the most up-to-date information.  We just need to know what the user called him-/herself the last time we saw him/her so that we can tell somebody else (i.e., a human).

Dave's point about the difference in cost between contacting the Keycloak broker vs contacting the Postgres server is important.  However, I'm not sure that the Keycloak _broker_ is going to know anything more than our Postgres server would, if the user changes his information on the OIDC identity _provider_ (assuming the user didn't go through our broker to get there).

Nevertheless, I don't think that we can realistically achieve a sufficiently authoritative bi-directional translation between UUID and username.  Given that the usernames are not unique, translation to UUID may produce multiple values, so that game is off.  And, translation the other way can fail for the same reason -- between the time that the request is made for the translation and the time it is responded to, the correspondence between UUIDs and usernames may have changed in such a way that we get an unintended result.

So, I really think that we need to limit the use of usernames to just presentation to human users.  Usernames can be used to search for things (datasets, users, et al.), but eventually the human user is going to have to point to something and say "this one", and the result of that needs to be an MD5, or a UUID, or something a lot less squishy than a "name".

Reply:
This gets, um, "really interesting", regardless of whether we're using an authoritative external Keycloak broker or an internal cache. The whole idea of having users from distinct authorities ...

Really it breaks down your concept of "We just need to know what the user called him-/herself the last time we saw him/her so that we can tell somebody else". Sure, we could cache the username associated with a UUID when that users pushes a dataset, in expectation of being able to reverse translate later; but that cache could still have multiple UUIDs associated with that same username.

The only way I can see to accomplish a unique reverse translation would be to actually store both the UUID and name of the user that *owns* (initially uploaded, or later inherits) a dataset; which also means it becomes impractical and potentially expensive to track subsequent changes in the username associated with that UUID. (And I don't think we'd want to use a potentially stale value to match a search, either.)

I don't like any of the options. Caching seems like both the best and the worst alternative, except that it's both better and worse than any other alternative... 🤣

Reply:
> The only way I can see to accomplish a unique reverse translation [...]

I think we need to let go of the idea that there is or will be a unique translation of a username to a UUID.  However, I think that in nearly every case we can get by without it.

> store both the UUID and name of the user that owns (initially uploaded, or later inherits) a dataset

I don't think we would need to store both in each dataset.  All we would have to do to indicate the owner of a dataset is store in it either the owner's UUID or a reference (i.e., a foreign key) to a database entry which contains it.  If we have the foreign key, then we can do table joins for username-based queries.  If we store the just UUID, then we can first do a cache lookup to get a list of UUIDs and do an or-query for datasets.  Either way, I think we can get a pertinent list of results.  We can then apply a filter to select the ones for which the requesting user's UUID has access.